### PR TITLE
[WIP] Native clipboard support

### PR DESCRIPTION
**Pull request by JaySandhu**
# Problem:

The current provider based clipboard integration, while being very flexible, has a few major disadvantages:
  1. It's slow. Creating processes is slow (especially on Windows and MacOS).
  2. Block pasting does not work correctly out of the box.
  3. It's yet another finicky configuration point.

# Proposal:

A simple clipboard API that is implemented for Windows, MacOS, X11, and Wayland. The proposed API consists of three functions:

1. Init -  Called early in the startup process, allows the clipboard subsystem an opportunity to initialize. This may not be needed on some platforms, in which case it's a no-op.

2. Get - Blocking synchronous access to the system clipboard. Returns a string and a register type. Internally, we use `str_to_register` to set the appropriate register.

3. Set - Sets the system clipboard. This is done in a way that preserves the register type for other Neovim and Vim processes (required for block pasting to work correctly).

The spellings and prototypes for each function are up for discussion. Feedback appreciated.

 **Project Layout:**
`os/clipboard.h` - API definitions.
`os/clipboard_macos.m` - Implementation for MacOS.
`os/clipboard_win.c` - Implementation for Windows.
`os/clipbaord_unix.c` - Implementation for X11 and Wayland.

We use the build system to conditionally compile each platforms implementation. A special note for Unix: One of Neovim's goals is to have single build for each platform, to support this we can include both an X11 and Wayland implementation on Linux.

This pull request includes an initial draft of the proposed API in `os/clipbaord.h` and an implementation for MacOS in `os/clipboard_macos.m`. Block pasting works across processes and is compatible with Vim (closing https://github.com/neovim/neovim/issues/1822).

The next easiest platform to get up and running is Windows, we can implement the clipboard API fairly easily using the win32 API. Linux will be more challenging, but it's doable.

Once we have basic clipboard support working across the platforms, we can look into reviving `clipboard=autoselect`.

### CODE REVIEW

File: src/nvim/CMakeLists.txt [hunk 1]
```diff
@@ -16,8 +16,8 @@ if(WIN32)
   # tell MinGW compiler to enable wmain
   set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -municode")
 elseif(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
-  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -framework CoreFoundation")
-  set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} -framework CoreFoundation")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -framework AppKit")
```
Comment (Shougo) [inline src/nvim/CMakeLists.txt:19]: Why it has extra indentation?

Comment (JaySandhu) [reply src/nvim/CMakeLists.txt]: My mistake, had the wrong indentation settings. Fixed it.

**Comment by JaySandhu** [comment]
I plan on implementing Windows support over the weekend. Are there any objections to the outlined approach? I'd hate to spend a bunch of time working on something that is unlikely to be merged.

**Comment by tjdevries** [comment]
I can't speak for everyone, but from what I've seen so far I really like it. I see :+1: from several of the other maintainers as well. I will ask and make sure that no one is opposed.

**Comment by jamessan** [comment]
I understand the impetus, since some systems have pretty poor performance around process creation.  The change to providers also changed the X11 model from pull (assert ownership, but don't do anything else until someone requests the data) to push (immediately push the data to the provider), which basically demotes the X11 behavior to the more typical clipboard behavior.

If the providers could be long-lived, rather than spawning a process every time, that would the process creation performance issue.  Unfortunately, I don't think there are already tools that we could leverage for that.

The issue of preserving block copy seems like something we should be able to address without needing a native clipboard implementation, but maybe I just haven't looked into it enough.

I think leaving the X11/Wayland side for last is a mistake since I sense it's going to be the most difficult.  There are still significant issues in Vim's X11 clipboard support (like the lack of support for large copy/pastes because [INCR property handling](https://tronche.com/gui/x/icccm/sec-2.html#s-2.5) isn't implemented, so Vim will likely crash).

I think it's also important to realize that the provider infrastructure has given us the ability to provide flexible copy/paste support, and easily let the user plugin other tools.  A prime example is that we leverage tmux's paste-buffer when we detect we're running inside it, so you can easily copy/paste between tmux panes (which I keep forgetting about).

**Comment by JaySandhu** [comment]
Block pasting can be fixed under the provider based system. It would involve implementing Vim aware (or more generally type aware) clipboard providers for the various platforms. It's a fair amount of work. 

Long lived providers could also work. But, as you noted, there are no existing tools, so we'd be on the hook for implementing them.

We could make the native clipboard provider opt-in. Maybe by adding a `native` flag to the `clipboard` option. That way we preserve the flexibility of the current system for those who need it.

I'll try and tackle X11 support next. There was work in that area before (https://github.com/neovim/neovim/pull/4523), so I'll start there.

If native X11 / Wayland support doesn't come to pass, would it be out of the question to include native MacOS and Windows support?

**Comment by JaySandhu** [comment]
I spent the day messing around with clipboards on Linux, here's what I learnt.

First off, I didn't realize that Vim doesn't have native Wayland support. So to get to feature parity with Vim, we only need native X11 support.

**X11**
I'm confident we can abstract away the async pull nature of X11 selections and provide a more conventional clipboard API. Native X11 selection support, with support for incremental data transfers, would be around 1-2k lines of code.

**Wayland**
The situation on Wayland is pretty dire, I don't think it's workable to add a native solution. At least not yet.

Wayland predicates clipboard access on user interaction. In order to access the clipboard you need to have received a Wayland input event. The idea is that this prevents arbitrary processes from snooping on your clipboard. The problem is we don't receive input events, the UI (or terminal UI) does, so there's no way for us to manipulate the clipboard. This is true for all command line applications. There's been some effort to add protocol extensions that would remedy this problem, but they're not widely implemented.

[wl-clipboard](https://github.com/bugaevc/wl-clipboard) gets around this limitation by displaying a small transparent surface that temporarily assumes keyboard focus (Wayland clients with keyboard focus can also access the clipboard).

Unlike X11, Wayland has a lot implementation diversity and this comes with its own set of challenges. For example, Wayland core has done away with the primary selection concept, but it's commonly supported via protocol extensions, so we would have to dynamically ascertain if our current environment supports them.

**Comment by Shatur95** [comment]
Sounds very cool. What is the state of it?

**Comment by justinmk** [comment]
Looks fairly isolated (good) so far but the block selection issue is so minor, and performance issue as well, that it pains me to spend any time on this (as well as future bug reports, debugging, ...). https://github.com/neovim/neovim/pull/12452#issuecomment-642924554 expresses my thoughts exactly. 

Performance for `clipboard=unnamed` and `:global/.../y` can be fixed by skipping clipboard "set" operation until (1) `FocusLost` event or (2) user performs clipboard "get" operation. Would you be willing to look into that...?



For reference:

- @bfredl started the X11 case here: https://github.com/neovim/neovim/pull/4523 
- autoselect is WIP here: https://github.com/neovim/neovim/pull/3708

**Comment by JaySandhu** [comment]
I stopped working on this a while ago because it seemed unlikely to get merged. For what it's worth, I've been using the native macOS clipboard for the past two months and it's been great. I'd still be willing to put in the effort to get this over the line, but I understand the reluctance in accepting a >3000 line patch to fix block pasting.

Delaying setting the clipboard would definitely help for pathological cases (like holding down "x"), but if I'm being honest, it seems like a low value add. Also, it would break clipboard managers since they won't have access to the intermediate clipboard history. 

Another thing to consider, the clipboard can be set by us while we're out of focus, this can happen if a timer fires or due to a remote client. I guess that's easy enough to handle, but it is more complexity.

**Comment by justinmk** [comment]
> Delaying setting the clipboard would definitely help for pathological cases (like holding down "x"), but if I'm being honest, it seems like a low value add.

No. Delaying the clipboard **set** operation until user does clipboard **get** makes  `clipboard=unnamed` fast for the common case.

> Another thing to consider, the clipboard can be set by us while we're out of focus, this can happen if a timer fires or due to a remote client. I guess that's easy enough to handle, but it is more complexity.

When out-of-focus the performance hack can be disabled (the slowness won't be noticed by a human).

**Pull request closed without merging**